File: backend/app/schemas/croisement.py
"""Schemas Pydantic pour Pollen et Croisement"""
from pydantic import BaseModel, ConfigDict
from typing import Optional
from datetime import date, datetime


# ─── Pollen ──────────────────────────────────────────────────────────────────

class PollenCreate(BaseModel):
    nom_pollen: str
    id_variete_source: Optional[int] = None
    pheno_source: Optional[str] = None
    reverse: bool = False
    date_collecte: Optional[date] = None
    quantite_initiale_g: Optional[float] = None
    quantite_restante_g: Optional[float] = None
    stockage: Optional[str] = None            # frigo | congelateur | ambiant
    date_peremption: Optional[date] = None    # si absent, calculé auto
    notes: Optional[str] = None


class PollenUpdate(BaseModel):
    nom_pollen: Optional[str] = None
    id_variete_source: Optional[int] = None
    pheno_source: Optional[str] = None
    reverse: Optional[bool] = None
    date_collecte: Optional[date] = None
    quantite_initiale_g: Optional[float] = None
    quantite_restante_g: Optional[float] = None
    stockage: Optional[str] = None
    date_peremption: Optional[date] = None
    actif: Optional[bool] = None
    notes: Optional[str] = None


class PollenRead(BaseModel):
    id_pollen: int
    nom_pollen: str
    id_variete_source: Optional[int] = None
    nom_variete_source: Optional[str] = None
    pheno_source: Optional[str] = None
    reverse: bool = False
    date_collecte: Optional[date] = None
    quantite_initiale_g: Optional[float] = None
    quantite_restante_g: Optional[float] = None
    stockage: Optional[str] = None
    date_peremption: Optional[date] = None
    actif: bool = True
    notes: Optional[str] = None
    created_at: Optional[datetime] = None
    # Flags calculés
    perime: bool = False
    epuise: bool = False

    model_config = ConfigDict(from_attributes=True)


# ─── Croisement ──────────────────────────────────────────────────────────────

class CroisementCreate(BaseModel):
    nom_croisement: str
    type_croisement: Optional[str] = None              # F1 | F2 | BX | S1 | IBL | polyhybrid

    # Mère
    id_variete_mere: Optional[int] = None
    pheno_mere: Optional[str] = None
    notes_mere: Optional[str] = None

    # Père (pollen OU variété directe)
    id_pollen: Optional[int] = None
    id_variete_pere: Optional[int] = None
    pheno_pere: Optional[str] = None
    pere_reverse: bool = False
    notes_pere: Optional[str] = None

    # Pollinisation
    date_pollinisation: Optional[date] = None
    methode: Optional[str] = None
    zone_pollinisee: Optional[str] = None
    quantite_pollen_utilisee_g: Optional[float] = None

    statut: str = "planifie"
    notes: Optional[str] = None


class CroisementUpdate(BaseModel):
    nom_croisement: Optional[str] = None
    type_croisement: Optional[str] = None
    id_variete_mere: Optional[int] = None
    pheno_mere: Optional[str] = None
    notes_mere: Optional[str] = None
    id_pollen: Optional[int] = None
    id_variete_pere: Optional[int] = None
    pheno_pere: Optional[str] = None
    pere_reverse: Optional[bool] = None
    notes_pere: Optional[str] = None
    date_pollinisation: Optional[date] = None
    methode: Optional[str] = None
    zone_pollinisee: Optional[str] = None
    quantite_pollen_utilisee_g: Optional[float] = None
    statut: Optional[str] = None
    notes: Optional[str] = None


class RecolteGrainesInput(BaseModel):
    """Payload de finalisation d'un croisement → passage au statut 'recolte'."""
    date_recolte_graines: date
    nb_graines: int
    qualite_graines: Optional[str] = None             # bonne | moyenne | immature
    poids_graines_g: Optional[float] = None
    # Toggles sorties auto
    creer_variete: bool = True
    creer_packgraine: bool = True


class CroisementRead(BaseModel):
    id_croisement: int
    nom_croisement: str
    type_croisement: Optional[str] = None

    # Mère
    id_variete_mere: Optional[int] = None
    nom_variete_mere: Optional[str] = None
    pheno_mere: Optional[str] = None
    notes_mere: Optional[str] = None

    # Père
    id_pollen: Optional[int] = None
    nom_pollen: Optional[str] = None
    id_variete_pere: Optional[int] = None
    nom_variete_pere: Optional[str] = None
    pheno_pere: Optional[str] = None
    pere_reverse: bool = False
    notes_pere: Optional[str] = None

    # Pollinisation
    date_pollinisation: Optional[date] = None
    methode: Optional[str] = None
    zone_pollinisee: Optional[str] = None
    quantite_pollen_utilisee_g: Optional[float] = None

    # Résultat
    date_recolte_graines: Optional[date] = None
    nb_graines: Optional[int] = None
    qualite_graines: Optional[str] = None
    poids_graines_g: Optional[float] = None

    # Sorties
    id_variete_resultat: Optional[int] = None
    id_packgraine_resultat: Optional[int] = None

    statut: str = "planifie"
    notes: Optional[str] = None
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/croisement.ts
import client from './client'

// --- Types ---

export type TypeCroisement = 'F1' | 'F2' | 'BX' | 'S1' | 'IBL' | 'polyhybrid'
export type StatutCroisement = 'planifie' | 'pollinise' | 'maturation' | 'recolte' | 'echec'
export type MethodePollinisation = 'plante_entiere' | 'branche_isolee' | 'pinceau'
export type QualiteGraines = 'bonne' | 'moyenne' | 'immature'
export type StockagePollen = 'frigo' | 'congelateur' | 'ambiant'

// --- Pollen ---

export interface Pollen {
  id_pollen: number
  nom_pollen: string
  id_variete_source?: number
  nom_variete_source?: string
  pheno_source?: string
  reverse: boolean
  date_collecte?: string
  quantite_initiale_g?: number
  quantite_restante_g?: number
  stockage?: StockagePollen
  date_peremption?: string
  actif: boolean
  notes?: string
  created_at?: string
  perime: boolean
  epuise: boolean
}

export interface PollenCreate {
  nom_pollen: string
  id_variete_source?: number
  pheno_source?: string
  reverse?: boolean
  date_collecte?: string
  quantite_initiale_g?: number
  quantite_restante_g?: number
  stockage?: StockagePollen
  date_peremption?: string
  notes?: string
}

export interface PollenUpdate extends Partial<PollenCreate> {
  actif?: boolean
}

// --- Croisement ---

export interface Croisement {
  id_croisement: number
  nom_croisement: string
  type_croisement?: TypeCroisement

  id_variete_mere?: number
  nom_variete_mere?: string
  pheno_mere?: string
  notes_mere?: string

  id_pollen?: number
  nom_pollen?: string
  id_variete_pere?: number
  nom_variete_pere?: string
  pheno_pere?: string
  pere_reverse: boolean
  notes_pere?: string

  date_pollinisation?: string
  methode?: MethodePollinisation
  zone_pollinisee?: string
  quantite_pollen_utilisee_g?: number

  date_recolte_graines?: string
  nb_graines?: number
  qualite_graines?: QualiteGraines
  poids_graines_g?: number

  id_variete_resultat?: number
  id_packgraine_resultat?: number

  statut: StatutCroisement
  notes?: string
  created_at?: string
}

export interface CroisementCreate {
  nom_croisement: string
  type_croisement?: TypeCroisement
  id_variete_mere?: number
  pheno_mere?: string
  notes_mere?: string
  id_pollen?: number
  id_variete_pere?: number
  pheno_pere?: string
  pere_reverse?: boolean
  notes_pere?: string
  date_pollinisation?: string
  methode?: MethodePollinisation
  zone_pollinisee?: string
  quantite_pollen_utilisee_g?: number
  statut?: StatutCroisement
  notes?: string
}

export interface CroisementUpdate extends Partial<CroisementCreate> {}

export interface RecolteGrainesInput {
  date_recolte_graines: string
  nb_graines: number
  qualite_graines?: QualiteGraines
  poids_graines_g?: number
  // Breeder
  id_breeder?: number
  nom_breeder_nouveau?: string
  // Variete
  creer_variete?: boolean
  nom_variete_resultat?: string
  id_variete_existante?: number   // si creer_variete=false → variété à lier
  // Pack + graines
  creer_packgraine?: boolean
  types_graines?: string
}

// --- API calls ---

export const croisementAPI = {
  // Croisements
  list: (statut?: StatutCroisement) =>
    client.get<Croisement[]>('/croisements/', { params: statut ? { statut } : undefined }),
  get: (id: number) => client.get<Croisement>(`/croisements/${id}`),
  create: (data: CroisementCreate) => client.post<Croisement>('/croisements/', data),
  update: (id: number, data: CroisementUpdate) => client.put<Croisement>(`/croisements/${id}`, data),
  delete: (id: number) => client.delete(`/croisements/${id}`),
  finaliserRecolte: (id: number, data: RecolteGrainesInput) =>
    client.post<Croisement>(`/croisements/${id}/recolte`, data),

  // Pollen
  listPollen: (actifOnly = false) =>
    client.get<Pollen[]>('/croisements/pollen', { params: { actif_only: actifOnly } }),
  getPollen: (id: number) => client.get<Pollen>(`/croisements/pollen/${id}`),
  createPollen: (data: PollenCreate) => client.post<Pollen>('/croisements/pollen', data),
  updatePollen: (id: number, data: PollenUpdate) =>
    client.put<Pollen>(`/croisements/pollen/${id}`, data),
  deletePollen: (id: number) => client.delete(`/croisements/pollen/${id}`),
}
